bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/test/java/AssemblyConfigTest.java

```diff
@@ -0,0 +1,111 @@
+   * dependencySet} by finding an {@code include} for {@code hbase-common} in the {@code provided}
+   * {@code scope}.
+   */
+  @Test
+  public void verifyBYOHadoopIncludes() throws Exception {
+    // If this fails, the provided dependencySet in assembly/shaded-byo-hadoop.xml should be updated
+    // to reflect the deps specified by TableMapReduceUtil.
+    Assert.assertEquals(
+        "The assembly descriptor `shaded-byo-hadoop.xml` must be in sync with TableMapReduceUtil dependency jars",
+        extractHBaseBaseArtifactDeps(),
+        extractAssemblyBaseArtifactIncludes(new File("assembly/shaded-byo-hadoop.xml")));
+  }
+
+  private static Set<String> extractAssemblyBaseArtifactIncludes(File file)
+      throws ParserConfigurationException, XPathExpressionException, IOException, SAXException {
+    Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(file);
+    XPathFactory xPathFactory = XPathFactory.newInstance();
+    XPathExpression depSetPath =
+        xPathFactory
+            .newXPath()
+            .compile(
+                "/assembly/dependencySets/dependencySet[scope='provided' and includes/include='org.apache.hbase:hbase-common']");
+
+    NodeList depSetNodes = (NodeList) depSetPath.evaluate(doc, XPathConstants.NODESET);
+    Assert.assertEquals(
+        "Expected single dependencySet with provided hbase-common", 1, depSetNodes.getLength());
+
+    Node depSetNode = depSetNodes.item(0);
+    depSetPath = xPathFactory.newXPath().compile("includes/include");
+    NodeList includeNodes = (NodeList) depSetPath.evaluate(depSetNode, XPathConstants.NODESET);
+
+    Set<String> actualIncludeIds = new HashSet<>();
+    for (int i = 0; i < includeNodes.getLength(); i++) {
+      String versionlessCoordinate = includeNodes.item(i).getTextContent();
+      String artifactId = versionlessCoordinate.split(":")[1];
+      actualIncludeIds.add(artifactId);
+    }
+
+    return actualIncludeIds;
+  }
+
+  /**
+   * Extract the required base artifact ids from HBase's classpath.
+   *
+   * <p>This uses a heuristic that converts jar names into artifact ids. This approach is fragile,
+   * but shouldnt need to be updated much.
+   */
+  private static Set<String> extractHBaseBaseArtifactDeps() throws IOException {
+    Configuration configuration = new Configuration(false);
+    TableMapReduceUtil.addHBaseDependencyJars(configuration);
+    String classpath = TableMapReduceUtil.buildDependencyClasspath(configuration);
+
+    Set<String> expectedIncludeIds = new HashSet<>();
+    for (String pathStr : classpath.split(":")) {
+
+      String filename = new File(pathStr).getName();
+      // extract the artifact base name from the filename, stripping version and extension
+      String artifactId =
+          filename.replaceAll(
+              "(.*)-\\d+\\.\\d+(\\.\\d+)?(?:-(?:incubating|alpha|beta)|\\.Final)?\\.jar$", "$1");
```

Comment:
please extract this regex into a variable

Reply:
done